apps/frontend/src/layout/PrivateRoute.tsx

```diff
@@ -1,11 +1,243 @@
+import { GlobalSpinner, NotFound } from "@/components";
+import { createOrgDbs } from "@/database";
+import {
+    OrganizationCountableEdge,
+    useViewerLazyQuery,
+} from "@/generated/graphql";
+import { handleRedirect } from "@/modules/auth/helper";
 import { useAuthToken } from "@/modules/auth/hooks/useAuthToken";
-import { Navigate } from "react-router-dom";
+import { useSession } from "@/modules/users/hooks/useSession";
+import { SessionViewer } from "@/types";
+import { omitTypename } from "@/utils";
+import { logDebug } from "@/utils/log";
+import { useEffect, useState } from "react";
+import { Navigate, useNavigate, useParams } from "react-router-dom";
 
 type Props = {
     children: React.ReactNode;
 };
 export const PrivateRoute = ({ children }: Props) => {
+    const [ready, setReady] = useState(false);
+    const [invalidUrl, setInvalidUrl] = useState(false);
     const { token } = useAuthToken();
+    const { organizationSlug, projectId } = useParams();
+    const { viewer, createSession, updateSession } = useSession();
+    const navigate = useNavigate();
+    const [fetchViewer, { loading }] = useViewerLazyQuery();
 
-    return <>{!token ? <Navigate to="/" /> : children}</>;
+    useEffect(() => {
+        let user: SessionViewer | null = viewer || null;
+        let viewerIsLatest = false;
+        let shouldRedirect = false;
+        let org: OrganizationCountableEdge | null = null;
+
+        const validateUrl = async () => {
+            if (!token) {
+                return;
+            }
+            logDebug("viewer: ", user);
+            if (!user) {
+                logDebug(
+                    "No viewer, checking local storage...",
+                    "PrivateRoute",
+                );
+                const getCachedViewer = localStorage.getItem("viewer");
```

Comment:
@AfeezGL I know this PR is still work in progress but the view should care where the viewer is from (session, localstorage or backend). We should delegate this to a service layer or a hook. This has two advantages;

1. It makes the block of code shorter.
2. We can easily reuse this logic anywhere we need a viewer.